### PR TITLE
always run `configure_linker` except for mir-opt tests

### LINKED ISSUE #120768: "x.py check" dirties the build cache for "x.py build"
To reproduce, run
```
./x.py test ui --test-args pat
./x.py check compiler/rustc
./x.py test ui --test-args pat
```
The last command should finish immediately and not do any work. This worked fine until a few weeks ago.

However, now, this leads to rebuilds:
```
$ ./x.py test ui --test-args pat --bless
Building bootstrap
    Finished dev [unoptimized] target(s) in 0.02s
Building stage0 library artifacts (x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.07s
Building compiler artifacts (stage0 -> stage1, x86_64-unknown-linux-gnu)
   Compiling psm v0.1.21
   Compiling rustc_llvm v0.0.0 (/home/r/src/rust/rustc/compiler/rustc_llvm)
   Compiling stacker v0.1.15
   Compiling rustc_data_structures v0.0.0 (/home/r/src/rust/rustc/compiler/rustc_data_structures)
   Compiling rustc_span v0.0.0 (/home/r/src/rust/rustc/compiler/rustc_span)
   Compiling rustc_abi v0.0.0 (/home/r/src/rust/rustc/compiler/rustc_abi)
   Compiling rustc_type_ir v0.0.0 (/home/r/src/rust/rustc/compiler/rustc_type_ir)
   Compiling rustc_feature v0.0.0 (/home/r/src/rust/rustc/compiler/rustc_feature)
   Compiling rustc_error_messages v0.0.0 (/home/r/src/rust/rustc/compiler/rustc_error_messages)
   Compiling rustc_target v0.0.0 (/home/r/src/rust/rustc/compiler/rustc_target)
   Compiling rustc_ast v0.0.0 (/home/r/src/rust/rustc/compiler/rustc_ast)
   Compiling rustc_next_trait_solver v0.0.0 (/home/r/src/rust/rustc/compiler/rustc_next_trait_solver)
```
I don't know what this "psm" crate it but it seems to be where the problem starts.

### PR DESCRIPTION
**Pull request by onur-ozkan**
`configure_linker` now runs consistently unless it's for mir-opt tests. Previously `!= "check"` condition was causing dirt in the cargo cache between runs of `x anything-but-not-check` and `x check`.
 
 Fixes #120768
 
 cc @saethlin

### CODE REVIEW

**Review by saethlin — APPROVED** [review]
I've tried this out and it seems to work.

I'd love to see a cleaner signature for `Builder::cargo` in the future, but I'm happy to see any fix for this so quickly. Thanks!

**Comment by onur-ozkan** [comment]
> I'd love to see a cleaner signature for Builder::cargo in the future

This is definetly on my to-do plans.

File: src/bootstrap/src/core/build_steps/test.rs [hunk 1]
```diff
@@ -3266,6 +3267,7 @@ impl Step for CodegenGCC {
                 SourceType::InTree,
                 target,
                 "run",
+                false,
```
Comment (RalfJung) [inline src/bootstrap/src/core/build_steps/test.rs:3270]: Random unexplained booleans make code hard to read. This should either be an enum with meaningful variant names, or it needs a comment explaining what this `false` means.

**Comment by onur-ozkan** [comment]
Made some refactoring to how we handle linker configuration in `Builder::cargo`. I think it looks okay overall, but the downside is that we have to configure the linker manually when needed. Any thoughts? @RalfJung @saethlin

**Comment by RalfJung** [comment]
Seems too easy now to just forget calling `configure_linker` -- it's what you need in most cases, after all.

**Comment by onur-ozkan** [comment]
> the downside is that we have to configure the linker manually when needed. Any thoughts?

We don't do this anymore. `Builder::cargo` is now private, and we use these new functions instead:  https://github.com/onur-ozkan/rust/blob/3c019de0febceeed9432813c8dfd39280b244d5c/src/bootstrap/src/core/builder.rs#L2240-L2264.

File: src/bootstrap/src/core/build_steps/compile.rs [hunk 2]
```diff
@@ -237,12 +238,26 @@ impl Step for Std {
...
+                SourceType::InTree,
+                target,
+                "check",
+            );
             cargo.rustflag("-Zalways-encode-mir");
             cargo.arg("--manifest-path").arg(builder.src.join("library/sysroot/Cargo.toml"));
             cargo
         } else {
-            let mut cargo = builder.cargo(compiler, Mode::Std, SourceType::InTree, target, "build");
+            let mut cargo = builder::Cargo::new_for_mir_opt_tests(
```
Comment (RalfJung) [inline src/bootstrap/src/core/build_steps/compile.rs:253]: This is in the `!self.is_for_mir_opt_tests` arm but calls `new_for_mir_opt_tests`? Why that?

**Comment by RalfJung** [comment]
The approach makes sense to me, sadly I can't confidently review bootstrap changes.

@rust-lang/bootstrap would be good to land this rather quickly, this is fixing an issue that currently extends the edit-compile-test cycle for some cases from <1s to >1min.

**Comment by RalfJung** [comment]
This is fixing a rustc development roadblock.
@bors p=1

**Pull request merged**